## NAME-PAPU SWAIN

### PROJECT-03(Building user-based recommendation model for Amazon)

DOMAIN-ENTERTAINMENT

OBJECTIVE-Some of the movies hadn’t been watched and therefore, are not rated by the users. Netflix would like to take this as an opportunity and build a machine learning recommendation algorithm which provides the ratings for each of the users.

In [1]:
import pandas as pd
import numpy as np

In [2]:
amazon=pd.read_csv("Amazon - Movies and TV Ratings.csv")

In [3]:
amazon

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0     A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1      AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2     A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3      AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4     A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   
...              ...     ...     ...     ...     ...     ...     ...     ...   
4843  A1IMQ9WMFYKWH5     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4844  A1KLIKPUF5E88I     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4845   A5HG6WFZLO10D     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4846  A3UU690TWXCG1X     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4847   AI4J762YI6S06     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

      Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0        NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1        NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2        NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3        NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4        NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
...      ...     ...  ...       ...       ...       ...       ...       ...   
4843     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4844     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4845     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4846     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4847     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

      Movie202  Movie203  Movie204  Movie205  Movie206  
0          NaN       NaN       NaN       NaN       NaN  
1          NaN       NaN       NaN       NaN       NaN  
2          NaN       NaN       NaN       NaN       NaN  
3          NaN       NaN       NaN       NaN       NaN  
4          NaN       NaN       NaN       NaN       NaN  
...        ...       ...       ...       ...       ...  
4843       NaN       NaN       NaN       NaN       5.0  
4844       NaN       NaN       NaN       NaN       5.0  
4845       NaN       NaN       NaN       NaN       5.0  
4846       NaN       NaN       NaN       NaN       5.0  
4847       NaN       NaN       NaN       NaN       5.0  

[4848 rows x 207 columns]

In [4]:
amazon.describe().T

count      mean       std  min   25%  50%  75%  max
Movie1      1.0  5.000000       NaN  5.0  5.00  5.0  5.0  5.0
Movie2      1.0  5.000000       NaN  5.0  5.00  5.0  5.0  5.0
Movie3      1.0  2.000000       NaN  2.0  2.00  2.0  2.0  2.0
Movie4      2.0  5.000000  0.000000  5.0  5.00  5.0  5.0  5.0
Movie5     29.0  4.103448  1.496301  1.0  4.00  5.0  5.0  5.0
...         ...       ...       ...  ...   ...  ...  ...  ...
Movie202    6.0  4.333333  1.632993  1.0  5.00  5.0  5.0  5.0
Movie203    1.0  3.000000       NaN  3.0  3.00  3.0  3.0  3.0
Movie204    8.0  4.375000  1.407886  1.0  4.75  5.0  5.0  5.0
Movie205   35.0  4.628571  0.910259  1.0  5.00  5.0  5.0  5.0
Movie206   13.0  4.923077  0.277350  4.0  5.00  5.0  5.0  5.0

[206 rows x 8 columns]

### Movie has maximum views

In [5]:
pd.DataFrame(amazon.describe().T["count"].sort_values(ascending=False)[:1])

count
Movie127  2313.0

In [6]:
pd.DataFrame(amazon.drop(columns=["user_id"]).count().sort_values(ascending=False)[:1])

0
Movie127  2313

##### Movie 127 has maximum views i.e. 2313

#### Movies have maximum ratings

In [7]:
pd.DataFrame(amazon.drop(columns=["user_id"]).sum().sort_values(ascending=False)[:1])

0
Movie127  9511.0

##### Movie 127 has maximum ratings i.e. 9511

### Average rating for each movie

In [8]:
amazon.describe().T["mean"].sort_values(ascending=False)

Movie1      5.0
Movie66     5.0
Movie76     5.0
Movie75     5.0
Movie74     5.0
           ... 
Movie58     1.0
Movie60     1.0
Movie154    1.0
Movie45     1.0
Movie144    1.0
Name: mean, Length: 206, dtype: float64

In [9]:
pd.DataFrame(amazon.describe().T["mean"].sort_values(ascending=False)[:5])

mean
Movie1    5.0
Movie66   5.0
Movie76   5.0
Movie75   5.0
Movie74   5.0

In [10]:
pd.DataFrame(amazon.drop(columns=["user_id"]).mean().sort_values(ascending=False)[:5])

0
Movie1   5.0
Movie66  5.0
Movie76  5.0
Movie75  5.0
Movie74  5.0

### Above are the top 5 movies with the maximum ratings according to Average rating for each movie.

### Top 5 movies with the least audience

In [11]:
pd.DataFrame(amazon.drop(columns=["user_id"]).count().sort_values(ascending=True)[:5])

0
Movie1    1
Movie71   1
Movie145  1
Movie69   1
Movie68   1

## Recommendation Model

In [12]:
from surprise import Reader,Dataset

In [13]:
amazon.columns

Index(['user_id', 'Movie1', 'Movie2', 'Movie3', 'Movie4', 'Movie5', 'Movie6',
       'Movie7', 'Movie8', 'Movie9',
       ...
       'Movie197', 'Movie198', 'Movie199', 'Movie200', 'Movie201', 'Movie202',
       'Movie203', 'Movie204', 'Movie205', 'Movie206'],
      dtype='object', length=207)

In [14]:
melt_amazon=amazon.melt(id_vars=amazon.columns[0],value_vars=amazon.columns[1:],var_name="Movie_name",value_name="Rating")
melt_amazon

user_id Movie_name  Rating
0       A3R5OBKS7OM2IR     Movie1     5.0
1        AH3QC2PC1VTGP     Movie1     NaN
2       A3LKP6WPMP9UKX     Movie1     NaN
3        AVIY68KEPQ5ZD     Movie1     NaN
4       A1CV1WROP5KTTW     Movie1     NaN
...                ...        ...     ...
998683  A1IMQ9WMFYKWH5   Movie206     5.0
998684  A1KLIKPUF5E88I   Movie206     5.0
998685   A5HG6WFZLO10D   Movie206     5.0
998686  A3UU690TWXCG1X   Movie206     5.0
998687   AI4J762YI6S06   Movie206     5.0

[998688 rows x 3 columns]

In [15]:
reader=Reader(rating_scale=(-1,10))

data=Dataset.load_from_df(melt_amazon.fillna(0),reader=reader)
data

In [16]:
from surprise.model_selection import train_test_split

In [17]:
train,test=train_test_split(data,train_size=0.75,random_state=25)

In [18]:
from surprise import SVD

In [19]:
algo=SVD()
algo.fit(train)

In [20]:
pred=algo.test(test)
pred

[Prediction(uid='A3RNW9QT7VAB7V', iid='Movie148', r_ui=0.0, est=-0.018484389570647512, details={'was_impossible': False}),
 Prediction(uid='A3AB4HQ9HYY7EJ', iid='Movie5', r_ui=0.0, est=0.010035042210752125, details={'was_impossible': False}),
 Prediction(uid='A1XKFLOE7YJQAR', iid='Movie42', r_ui=0.0, est=0.0013045904589365463, details={'was_impossible': False}),
 Prediction(uid='A1ICAAD0883C47', iid='Movie82', r_ui=0.0, est=0.007067742856064681, details={'was_impossible': False}),
 Prediction(uid='A2USNS57VQTNEV', iid='Movie160', r_ui=0.0, est=-0.029865739953772022, details={'was_impossible': False}),
 Prediction(uid='A2OVLJUYCHMQ9L', iid='Movie150', r_ui=0.0, est=0.0005650966853389576, details={'was_impossible': False}),
 Prediction(uid='A289HZXEMAX65J', iid='Movie103', r_ui=0.0, est=-0.5406842834270826, details={'was_impossible': False}),
 Prediction(uid='AGXNUVOH83ZQL', iid='Movie80', r_ui=0.0, est=-0.00073467547795959, details={'was_impossible': False}),
 Prediction(uid='A316HU3BAF

In [21]:
from surprise import accuracy

In [22]:
accuracy.rmse(pred)

RMSE: 0.2793


0.2792605553741894

In [23]:
accuracy.mae(pred)

MAE:  0.0415


0.04152275574001286

In [24]:
from surprise.model_selection import cross_validate

In [25]:
cv=cross_validate(algo,data,measures=["rmse","mae"],cv=3,verbose=True)
cv

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.2784  0.2854  0.2833  0.2824  0.0029  
MAE (testset)     0.0427  0.0432  0.0428  0.0429  0.0002  
Fit time          53.72   54.00   53.74   53.82   0.13    
Test time         4.26    4.18    4.98    4.47    0.36    


{'test_rmse': array([0.27843741, 0.28535171, 0.28326526]),
 'test_mae': array([0.04273999, 0.04320125, 0.04278652]),
 'fit_time': (53.71716785430908, 54.00098013877869, 53.74315047264099),
 'test_time': (4.263526201248169, 4.17860221862793, 4.9811179637908936)}

In [26]:
def repeat(algo_type,frame,min_,max_):
    reader=Reader(rating_scale=(-1,10))
   
    data=Dataset.load_from_df(frame,reader=reader)
    
    print(cross_validate(algo_type,data,measures=["RMSE","MAE"],cv=3,verbose=True))
    
    print("#"*10)
    user_id="A3R5OBKS7OM2IR"
    movie_id="Movie1"
    r_ui= 5.0
    print(algo_type.predict(user_id,movie_id,r_ui=r_ui,verbose = True))
    
    print("#"*10)

In [27]:
repeat(SVD(),melt_amazon.fillna(melt_amazon["Rating"].mean()),-1,10)

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.0832  0.0872  0.0873  0.0859  0.0019  
MAE (testset)     0.0098  0.0100  0.0097  0.0098  0.0001  
Fit time          54.25   54.91   55.18   54.78   0.39    
Test time         4.37    4.41    4.92    4.57    0.25    
{'test_rmse': array([0.08319674, 0.087217  , 0.08730632]), 'test_mae': array([0.0098303 , 0.00995434, 0.00974673]), 'fit_time': (54.253835916519165, 54.91045594215393, 55.17532539367676), 'test_time': (4.369490623474121, 4.411482572555542, 4.916153430938721)}
##########
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 4.38   {'was_impossible': False}
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 4.38   {'was_impossible': False}
##########


In [28]:
repeat(SVD(),melt_amazon.fillna(melt_amazon["Rating"].median()),-1,10)

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.0902  0.0918  0.0963  0.0927  0.0026  
MAE (testset)     0.0093  0.0086  0.0088  0.0089  0.0003  
Fit time          57.24   58.57   56.18   57.33   0.98    
Test time         4.40    4.40    4.34    4.38    0.03    
{'test_rmse': array([0.09016304, 0.09175172, 0.09626035]), 'test_mae': array([0.00926776, 0.00855696, 0.00875794]), 'fit_time': (57.24009609222412, 58.57158398628235, 56.177730560302734), 'test_time': (4.395478248596191, 4.401470899581909, 4.339505910873413)}
##########
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 5.00   {'was_impossible': False}
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 5.00   {'was_impossible': False}
##########


In [29]:
from surprise.model_selection import GridSearchCV

In [30]:
parameter_grid={"n_epochs":[20,30],"lr_all":[0.005,0.01],"n_factors":[50,100]}
parameter_grid

{'n_epochs': [20, 30], 'lr_all': [0.005, 0.01], 'n_factors': [50, 100]}

In [31]:
gs=GridSearchCV(SVD,parameter_grid,measures=["rmse","mae"],cv=3)

data1=Dataset.load_from_df(melt_amazon.fillna(melt_amazon["Rating"].mean()),reader=reader)

gs.fit(data1)

In [32]:
gs.best_score

{'rmse': 0.08504778116505778, 'mae': 0.008519749531631407}

In [33]:
print(gs.best_params["rmse"])
print(gs.best_score["rmse"])

{'n_epochs': 30, 'lr_all': 0.01, 'n_factors': 50}
0.08504778116505778


#                                THANK YOU!!!